dbuild/main.d

```diff
@@ -0,0 +1,311 @@
+core/sync/semaphore.di
+  `.splitEntries.array;
+
+/+
+Returns whether to include `file` in `DOCS`
++/
+bool isValid_DOCS(string file)
+{
+    file = file.stripExtension.replace("/", ".");
+    import std.stdio;
+
+    if (file.startsWith("rt.", "core.stdc.", "core.stdcpp.", "core.sync."))
+        return true;
+
+    import std.regex;
+
+    if (matchFirst(file, `^(object|core\.\w+)$`.regex))
+        return true;
+    return false;
+}
+
+void main(string[] args)
+{
+    auto builder = new Builder;
+    builder.generateMakefileVariables(args);
+}
+
+/+
+this auto-generates Makefile variables
+for SRCS,DOCS,COPY, it scans files in source directories and applies exclusion
+rules: see isValid_{SRCS,DOCS,COPY}.
++/
+class Builder
+{
+
+    // Verifies auto-generated `files` match existing `filesGold`
+    void verify(string name, string[] files, string[] filesGold)
+    {
+        auto temp1 = filesGold.sort().array;
+        auto temp2 = files.sort().array;
+
+        // PRTEMP: for debuggging (remove beofore submission)
+        {
+          temp1.join("\n").writeTo("/tmp/z02.temp1.txt");
+          temp2.join("\n").writeTo("/tmp/z02.temp2.txt");
+        }
+
+        auto temp12 = temp1.setDifference(temp2).array;
+        auto temp21 = temp2.setDifference(temp1).array;
+        writeln2(name, temp1.length, temp2.length, temp12.length, temp21.length, temp1 == temp2);
+        writeln2(temp12.join("\n"));
+        writeln2(temp21.join("\n"));
+        assert(temp1 == temp2);
+    }
+
+    // Verifies auto-generated `entries` match Makefile variable defined in `goldFile`
+    void verify2(string[] entries, string goldFile)
+    {
+        auto gold = getGold(goldFile);
+        verify(goldFile, entries, gold.array);
+    }
+
+    // reads and normalizes files defined by Makefile variable defined in `goldFile`
+    auto getGold(string goldFile)
+    {
+        return goldFile.readText.splitLines[1 .. $].map!(a => a.strip.replace(` \`,
+                ``)).filter!(a => !a.empty && a != `\`).map!(a => a.replace(`\`,
+                `/`)).array.sort.release;
+
+    }
+
+    /+
+    generate Makefile variables (see list in D20180211T125334)
+    +/
+    void generateMakefileVariables(string[] args)
+    {
+        writeln2(args);
+
+        // read environment
+        auto aa = environment.toAA;
+        auto outDir = aa["ROOT_OF_THEM_ALL"];
+        auto TOOL_RDMD = aa["TOOL_RDMD"];
+        auto GENERATED_VARS_F = aa["GENERATED_VARS_F"];
+
+        // define some Makefile variables
+        version (OSX)
+        {
+            string OS = "osx";
+        }
+        else version (linux)
+        {
+            string OS = "linux";
+        }
+
+        auto IMPDIR = `import`;
+        auto DOCDIR = `doc`;
+        auto srcDir = "src";
+
+        // list D files in `dir` (without `srcDir` path prefix)
+        auto getFiles(string dir)
+        {
+            // dfmt off
+            return dirEntries(srcDir.buildPath(dir), "*.d", SpanMode.depth)
```

Comment:
Generating file lists to build by globbing is a flawed principle. Any temporary, backup, test, or otherwise unwanted files are going to be picked up, and can mess up the build in unpredictable ways.

Ideally, there would be exactly one list of files to build, which is the source of truth for the various build steps.

Reply:
these are further filtered, see `isValid_DOCS, isValid_SRCS, isValid_COPY`; unwanted files/ subdirectories and any special rule would be included / excluded there.

The current status quo is not DRY and too error prone (eg see https://github.com/dlang/druntime/commit/7827ee4302a9976e9bd6216607925bd08df187aa)